unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+#![allow(unused, dead_code)]
+
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::fmt;
+use crate::pb::{FlatUnixFs, PBLink, PBNode, ParsingFailed, UnixFsType};
+use crate::file::{FileMetadata, FileReadFailed};
+use crate::file::visit::{IdleFileVisit, FileVisit, Cache};
+use crate::InvalidCidInLink;
+use std::path::{Path, PathBuf};
+use cid::Cid;
+
+#[derive(Debug)]
+pub struct Walker {
+    current: InnerEntry,
+    /// On the next call to `continue_walk` this will be the block, unless we have an ongoing file
+    /// walk in which case we shortcircuit to continue it.
+    next: Option<(Cid, String, usize)>,
+    pending: Vec<(Cid, String, usize)>,
+    // tried to recycle the names but that was consistently at fast and used more memory than just
+    // cloning the strings
+}
+
+fn convert_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let name = match link.Name {
+        Some(Cow::Borrowed(s)) if !s.is_empty() => s.to_owned(),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => s,
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+fn convert_sharded_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
```

Comment:
the magic `2` is dependent on ShardedLookup validating the fanout and hash_type beforehand, which has not been implemented.

Reply:
Leaving this unresolved to get back to this later on.

Reply:
Done already.